libbpfgo/libbpfgo.go

```diff
@@ -648,39 +650,71 @@ func (m *Module) InitRingBuf(mapName string, eventsChan chan []byte) (*RingBuffe
 }
 
 func (rb *RingBuffer) Start() {
+	rb.stop = make(chan struct{})
+	rb.wg.Add(1)
 	go rb.poll()
-	rb.stopped = false
 }
 
 func (rb *RingBuffer) Stop() {
-	rb.stopped = true
+	if rb.stop != nil {
+		// Tell the poll goroutine that it's time to exit
+		close(rb.stop)
+
+		// The event channel should be drained here since the consumer
+		// may have stopped at this point. Failure to drain it will
+		// result in a deadlock: the channel will fill up and the poll
+		// goroutine will block in the callback.
+		eventChan := eventChannels[uintptr(rb.bpfMap.fd)]
```

Comment:
Ah this was the pattern that I failed to recognize the need for in previous iterations of this code. I couldn't get past the fact that the callback was blocking (and the timeout passed to ring_buffer__poll would interrupt it).